### DISCUSSION TITLE
Multiple docker instances - Not booting issue

### DISCUSSION THREAD

**Discussion by SRoy89** [photoprism/photoprism#2358] [Q&A] [answered]
I sucessfully installed photoprism, started adding photos then realised there can't be multiple users for my family. As photoprism doesn't support multi-users, I thought of adding multi-containers of photoprism and mariadb with their own ports. But it doesn't boot.

I've changed ports to `2345` & `2346` and added `sroy` & `croy` respectively at the end of each container name.
I also had to add different ports to `mariadbsroy` & `mariadbcroy` containers, else both photoprism containers talks with the same db container (`3305` & `3306` respectively).

I also checked [logs](https://i.imgur.com/dDEGGrP.jpeg) which are fine. I have 4GB Swap. As for the volumes:
<img src="https://user-images.githubusercontent.com/61577078/170495978-c2c5eda2-8c8e-49f5-92ab-6c0171f0b58b.png" width="300" />

Normally it boots on external hdd, so I probably have messed something trying to do multi instances.  I hope to be able to add more users, when I can figure out at least two users.

Here's the link to the docker-compose.yaml file: [https://pastebin.com/wNRLwcUi](https://pastebin.com/wNRLwcUi).

Thanks

**Comment by heikomat** [comment, marked as answer]
You told docker to use different ports, but you didn't tell photoprism to use different ports.

When you configured this:
```
ports:
      - "2345:2345" # HTTP port (host:container)
```
You basically told docker: "Whatever is running inside the container on port `2345` (the second number), make it available on the host on port `2345` (the first number)".

Photoprism by default runs on port `2342`, and you haven't told photoprism to run on a different port. That means inside your container, port `2345` is unused.

You can use the [`PHOTOPRISM_HTTP_PORT`](https://docs.photoprism.app/getting-started/config-options/#web-server)-Environment-Variable to tell photoprism what port to use. Alternatively you could forward port `2345:2342` (for sroy) and `2346:2342` (for croy)

**Reply by SRoy89** [reply]
> Alternatively you could forward port 2345:2342 (for sroy) and 2346:2342 (for croy)

I see. I added `PHOTOPRISM_HTTP_PORT` to both. I thought of port forwarding before, but I assumed internal ports would conflict. 

After testing I can confirm that I was right. Photoprism is actually communicating to MariaDB via its internal port.
> PHOTOPRISM_DATABASE_SERVER: "mariadbcroy:3306"     # MariaDB or MySQL database server (hostname:port)

>mariadbcroy:
    ports:
      - "3307:3306"

I've tried to replace db server with `localhost:3007` & `http://localhost:3007` & `http://192.168.0.175:3007`.

None of them worked, but when I did `PHOTOPRISM_DATABASE_SERVER: "mariadbcroy:3306"` it worked because it **only** accepts internal host:port.
Doesn't matter what I write in the external port of `mariadbcroy:` as long as the internal is `3006` it works, as it's the default port.


So `mariadbsroy` can't use the same internal port. **So is there a way to change the default internal MariaDB port?**
 

TL;DR: If I what I wrote is really confusing here's the [current docker-compose.yaml in pastebin](https://pastebin.com/TCaS0cm2), just know in current settings sroy doesn't work, but croy does work because photoprism only accepts mariadb's internal host & port.

**Reply by heikomat** [reply]
> > Alternatively you could forward port 2345:2342 (for sroy) and 2346:2342 (for croy)
> 
> I see. I added `PHOTOPRISM_HTTP_PORT` to both. I thought of port forwarding before, but I assumed internal ports would conflict.
> 
> After testing I can confirm that I was right. Photoprism is actually communicating to MariaDB via its internal port.
> 
> > PHOTOPRISM_DATABASE_SERVER: "mariadbcroy:3306"     # MariaDB or MySQL database server (hostname:port)
> 
> > mariadbcroy:
> > ports:
> > - "3307:3306"
> 
> I've tried to replace db server with `localhost:3007` & `http://localhost:3007` & `http://192.168.0.175:3007`.
> 
> None of them worked, but when I did `PHOTOPRISM_DATABASE_SERVER: "mariadbcroy:3306"` it worked because it **only** accepts internal host:port. Doesn't matter what I write in the external port of `mariadbcroy:` as long as the internal is `3006` it works, as it's the default port.
> 
> So `mariadbsroy` can't use the same internal port. **So is there a way to change the default internal MariaDB port?**
> 
> TL;DR: If I what I wrote is really confusing here's the [current docker-compose.yaml in pastebin](https://pastebin.com/TCaS0cm2), just know in current settings sroy doesn't work, but croy does work because photoprism only accepts mariadb's internal host & port.

First of: there is a typo in your config. you set the following environment-variable for sroy:
```
PHOTOPRISM_DATABASE_SERVER: "mariadbsroy:330s"
```
Note the `s` instead of the `5`. Maybe that's all there is to it. If you still need to change mariaDBs internal port, you can read [here](https://hub.docker.com/r/arm64v8/mariadb/#!) how to configure the mariadb-container.
The easiest would probably be to add `--port [somePort]` to the `command`.

**Reply by SRoy89** [reply]
> Note the s instead of the 5. Maybe that's all there is to it.

Oops, corrected it. But that didn't fix it.

> If you still need to change mariaDBs internal port, you can read [here](https://hub.docker.com/r/arm64v8/mariadb/#!) how to configure the mariadb-container.
The easiest would probably be to add --port [somePort] to the command.

The only method of changing port [there](https://hub.docker.com/r/arm64v8/mariadb/#!) was the same one you suggested which is `--port 3305` to `command:` in `mariadbsroy:`. But that also didn't work. Moreover, adding that port command made `photoprismcroy` not to work which was weird as I didn't change a single thing in `croy`, I double checked.

I found [this reply](https://stackoverflow.com/a/56213000/14120176) on StackOverflow, which suggests that the container image could be "statically" bound to `:3306`. True, that image and this image are different, but it could be a possibility, and I also don't have a `.cnf` file. Also, the solution provided there only changes the external port which doesn't work as it **only** accepts the internal host:port.

This is getting kinda complicated, I was hoping to have a simple solution to this problem. One database server should just be able to handle multiple clients like `sroy`, `croy`, but for the sake of simplicity, I made a copy of the service. I can't imagine how complex it'll get to make multiple accounts work. I am beginning to feel like I should just wait for the devs to finish adding support for multi accounts.

I still am looking for the solution, as I decided to make the switch from google photos to being independently self-hosted.

Edit: I also tried adding `--port 3305` and `--port 3306` to both, which also didn't work.

**Reply by heikomat** [reply]
There has to be something wrong in your docker compose.
It is definitely possible to run multiple photoprism instances on one host. I have 4 instances on one host myself, just without docker compose (running the required containers via systemd)

I can take a closer look and maybe test your docker compose myself when I'm back home in about 2:30h

**Reply by SRoy89** [reply]
> There has to be something wrong in your docker compose. It is definitely possible to run multiple photoprism instances on one host. I have 4 instances on one host myself, just without docker compose (running the required containers via systemd)
> 
> I can take a closer look and maybe test your docker compose myself when I'm back home in about 2:30h

Thanks but there's no need!

**I got it to work, all 3 of them!** The issue was that because `croy` was working fine without the `PHOTOPRISM_HTTP_PORT: `,  I didn't add it when testing `sroy`. But now adding `--port` command to both DBs plus adding `PHOTOPRISM_HTTP_PORT:` to both photoprism, it worked.
Thanks a lot @heikomat, you helped me a lot!

Current and Working docker-compose file: [https://pastebin.com/JWHH1J2V](https://pastebin.com/JWHH1J2V)


**One last question is there a way to change the usernames from `Admin` to sroy, croy, and droy?** I searched the docs but couldn't find anything on it. (perhaps I should make a different topic on that and close this topic)


![image](https://user-images.githubusercontent.com/61577078/170821459-e1a30a39-60cc-422d-ba6b-ae0eeabe3b6a.png)

**Comment by rickysarraf** [comment]
I can confirm too that it must be a problem with your setup. I have 2
photoprism instances running on the same host

s3nt fr0m a $martph0ne, excuse typ0s

**Reply by SRoy89** [reply]
There were indeed problems with my setup, it's because I was testing a combination of multiple and it was hard to keep track of what I did and didn't test. Thanks for pointing it out, made me recheck everything.
